distributed_loader.cc

```diff
@@ -450,6 +451,42 @@ distributed_loader::process_upload_dir(distributed<database>& db, distributed<db
     });
 }
 
+future<std::tuple<utils::UUID, std::vector<std::vector<sstables::shared_sstable>>>>
+distributed_loader::get_sstables_from_upload_dir(distributed<database>& db, sstring ks, sstring cf) {
+    seastar::thread_attributes attr;
+    attr.sched_group = db.local().get_streaming_scheduling_group();
+    return seastar::async(std::move(attr), [&db, ks = std::move(ks), cf = std::move(cf)] {
+        global_column_family_ptr global_table(db, ks, cf);
+        sharded<sstables::sstable_directory> directory;
+        auto table_id = global_table->schema()->id();
+        auto upload = fs::path(global_table->dir()) / "upload";
+
+        directory.start(upload, db.local().get_config().initial_sstable_loading_concurrency(), std::ref(db.local().get_sharded_sst_dir_semaphore()),
+            sstables::sstable_directory::need_mutate_level::yes,
+            sstables::sstable_directory::lack_of_toc_fatal::no,
+            sstables::sstable_directory::enable_dangerous_direct_import_of_cassandra_counters(db.local().get_config().enable_dangerous_direct_import_of_cassandra_counters()),
+            sstables::sstable_directory::allow_loading_materialized_view::no,
+            sstables::sstable_directory::load_and_stream::yes,
+            [&global_table] (fs::path dir, int64_t gen, sstables::sstable_version_types v, sstables::sstable_format_types f) {
+                return global_table->make_sstable(dir.native(), gen, v, f, &error_handler_gen_for_upload_dir);
+
+        }).get();
+
+        auto stop = defer([&directory] {
+            directory.stop().get();
+        });
+
+        std::vector<std::vector<sstables::shared_sstable>> sstables_on_shards(smp::count);
+        lock_table(directory, db, ks, cf).get();
+        process_sstable_dir(directory).get();
+        directory.invoke_on_all([&sstables_on_shards] (sstables::sstable_directory& d) mutable {
+            sstables_on_shards[this_shard_id()] = d.get_sstables_load_and_stream();
+        }).get();
+
+        return std::make_tuple(table_id, sstables_on_shards);
+    });
+}
```

Comment:
Consider using a coroutine instead of a thread.

Reply:
I will try to have some fun.